atomic_reactor/util.py

```diff
@@ -1557,3 +1558,49 @@ def has_operator_manifest(workflow):
     except KeyError:
         operator_label = 'false'
     return operator_label.lower() == 'true'
+
+
+def annotation(*keys):
+    """
+    Annotate a `BuildPlugin` subclass. The `run()` method of this plugin will
+    store its result in the plugin's workflow under the given keys.
+
+    Example:
+    >>> from atomic_reactor.plugin import BuildPlugin
+
+    >>> @annotation('foo', 'bar')
+    >>> class MyBuildPlugin(BuildPlugin):
+    >>>     key = 'my_build_plugin'
+    >>>
+    >>>     def run(self):
+    >>>         return 1
+
+    >>> p = MyBuildPlugin(...)
+    >>> p.run()
+    >>> p.workflow.annotations  # {'foo': 1, 'bar': 1}
+
+    :param keys: Keys to annotate the plugin with
+    :return: Decorator that will turn the plugin into an annotated one
+    """
+    from atomic_reactor.plugin import BuildPlugin
+
+    def annotate(cls):
+        if not issubclass(cls, BuildPlugin):
+            raise TypeError('Only subclasses of BuildPlugin can be annotated')
+
+        run = cls.run
+
+        @functools.wraps(run)
+        def annotated_run(self):
+            result = run(self)
+            annotations = self.workflow.annotations
+            for key in keys:
+                if key in annotations:
+                    raise RuntimeError('Annotation {!r} is already set'.format(key))
+                annotations[key] = result
```

Comment:
The reason for wanting to specify multiple annotation keys is to be able to return multiple *different* values and assign them to different keys. For example, we might want to set both 'pullspecs':'example.registry.com/foo/bar:baz' and 'digests':{'v1':'spam','v2':'maps'} from the same plugin, and we'd want the annotations to be:
* 'pullspecs', with the string value 'example.registry.com/foo/bar:baz'
* 'digests', with the string value being the JSON encoding of the dict, i.e. '{"v1": "spam", "v2": "maps": }'

So there are several ways to get the different values assigned to the keys:
* maybe the run() method has to return a tuple, and we map each tuple item to the keys in the order they were given (raise an error if the numbers don't match?)
* or maybe run() has to return a dict, and the values are fetched from that dict by the annotation key names provided
* or maybe you can think of a nicer scheme for it